File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        if (head == nullptr) {
            return false;
        }
        ListNode* fastPtr = head;
        ListNode* slowPtr = head;
        while (fastPtr->next != nullptr) {
            fastPtr = fastPtr->next->next;
            if (fastPtr == nullptr) {
                return false;
            }
            slowPtr = slowPtr->next;
            if (slowPtr == fastPtr) {
                return true;
            }
        }
        return false;
    }
};

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        return reverseHelper(null, head);
    }
    ListNode reverseHelper(ListNode prev, ListNode curr) {
        if (curr == null) return prev;  // prev is new head
        ListNode next = curr.next;
        curr.next = prev;
        return reverseHelper(curr, next);
    }
}
